Flag Statistics Sub-Command

Bentleysb opened the pull request:
### Description:

This adds a command for counting flags in log files on mass. 
Given a directory of ISO country folders containing log files, it will get the number of flags per check per country. The result is a table of the collected data. The data is written to a csv file(s) in the provided output folder. 
Additionally, this command can take in two directories as inputs and calculate the difference between the flag counts. 
This command is built on top of the [Atlas Shell Tools](https://github.com/osmlab/atlas/tree/dev/atlas-shell-tools) system from Atlas. 

There are 3 output types: full, totals, and counts
By default the full output is used, but any of them can be enabled or disabled using the `--output-type` argument.

Example full output:

|Check|ABC|XYZ|Total|
|---|---|---|---|
|Check1|6|8|14|
|Check2|2||2|
|Check3||2|2|
|Total|8|10|18|

Example full difference output:

|Check|ABC|XYZ|Total|
|---|---|---|---|
|Check1|0|4|4|
|Check2|-2||-2|
|Check3||0|0|
|Total|-2|4|2|

Example totals output with reference and difference columns:

|Check|Reference|Input|Difference|
|---|---|---|---|
|Check1|10|14|4|
|Check2|4|2|-2|
|Check3|2|2|0|

Example counts output with reference and difference columns:

|Country|Check|Reference|Input|Difference|
|---|---|---|---|---|
|ABC|Check1|6|6|0|
|ABC|Check2|4|2|-2|
|ABC|Check3||||
|XYZ|Check1|4|8|4|
|XYZ|Check2||||
|XYZ|Check3|2|2|0|



### Potential Impact:

None

### Unit Test Approach:
Unit test check that both the count and diff functions work as expected.

### Test Results:

Tested on atlas checks runs of over 200 countries. A sampling of returned counts and diffs were as expected.

Bentleysb commented on src/main/java/org/openstreetmap/atlas/checks/commands/FlagStatisticsSubCommand.java:
```diff
@@ -0,0 +1,330 @@
+public class FlagStatisticsSubCommand extends AbstractAtlasShellToolsCommand
+{
+    private static final String INPUT_OPTION = "input";
+    private static final String DIFFERENCE_OPTION = "difference";
```
Reference does not make a lot of sense, as this is the target of the diff, not the source.

danielduhh replied:
The source is `input`, and the target (or reference) is what you're comparing this to, right? I think reference makes sense...

Bentleysb commented on src/main/java/org/openstreetmap/atlas/checks/commands/FlagStatisticsSubCommand.java:
```diff
@@ -0,0 +1,330 @@
+import com.google.gson.JsonObject;
+
+/**
+ * This command takes a folder of directories of atlas-checks log files and counts the number of
```
It does not have to be that flag folder, though that is the expected input most of the time.

Bentleysb commented on src/main/java/org/openstreetmap/atlas/checks/commands/FlagStatisticsSubCommand.java:
```diff
@@ -0,0 +1,330 @@
+                        .equalsIgnoreCase("log"))
+                .forEach(file ->
+                {
+                    logger.info("Reading: {}", file.getName());
```
Not necessarily, but it makes it so you arn't looking at a blank screen for 3 minutes.

Bentleysb commented:
@danielduhh, that run time sounds about right for a diff of 2 full atlas-checks runs. It just takes a while to read in the thousands of flags.

Bentleysb commented:
@danielduhh, I have added the two requested outputs. They can be enabled using the `--output-types` argument. There are example in the PR description. Additionally, the original output is now generated in three files, if a difference is requested. One file gives the input counts, one the reference counts, and the third the difference counts. To accommodate this the output parameter now takes a folder instead of a file. 
I also changed to using parallelStreams and ConcurrentHashMaps for reading the log files. This has decreased the overall runtime by about 200%.

danielduhh commented:
Nice job optimizing the run time.
What's the difference between the `total` and `counts` output? In your example, it looks like one has the country field, one does not. If that's the case, a single output with `country,check,source,reference,diff` should work

danielduhh commented on src/main/java/org/openstreetmap/atlas/checks/commands/FlagStatisticsSubCommand.java:
```diff
@@ -0,0 +1,619 @@
+                    .forEach((check, count) -> difference.get(country).put(check,
+                            new Counter(target.getOrDefault(country, new HashMap<>())
+                                    .getOrDefault(check, new Counter()).getValue()
+                                    - count.getValue())));
```
I think this is backwards. If the `target` (or baseline) is 20, and the `source` (or input) is 10, the result should be -10.

Bentleysb replied:
I think this gets back to terminology again. The input is the source/baseline, the reference is the target. So input/source/baseline should be older than reference/target.

danielduhh replied:
Your example shows -2 as the difference between 4 (input) & 2 (reference); However, the new run (input) has 2 additional flags from the comparison, so it should be +2, right?
```csv
Country,Check,Input,Reference,Difference
ABC,Check2,4,2,-2
```

Bentleysb replied:
The new run is reference.

Bentleysb commented:
@danielduhh, counts is countrified, totals is not (so it is the total of all countries for each check).

Bentleysb commented:
Based on offline discussion, I have updated the reference input to be treated as the older data set.

danielduhh commented on src/main/java/org/openstreetmap/atlas/checks/commands/FlagStatisticsSubCommand.java:
```diff
@@ -0,0 +1,619 @@
+            // Write the totals output if requested
+            if (outputTypes.contains("totals"))
+            {
+                this.writeCSV(outputFolder + "/totals.csv", totalsOutput);
```
totals.csv -> checkSummary.csv. Also, can you break this out into a constant

danielduhh commented on src/main/java/org/openstreetmap/atlas/checks/commands/FlagStatisticsSubCommand.java:
```diff
@@ -0,0 +1,619 @@
+                // Write outputs for the reference and difference if requested
+                if (outputTypes.contains("full"))
+                {
+                    this.writeCSV(outputFolder + "/reference.csv",
```
I don't think this output is necessary. This will be a duplicate of the "input.csv" output if we continuously output each `output-types` from run to run

danielduhh commented on src/main/java/org/openstreetmap/atlas/checks/commands/FlagStatisticsSubCommand.java:
```diff
@@ -0,0 +1,619 @@
+                {
+                    this.writeCSV(outputFolder + "/reference.csv",
+                            generateFullOutput(referenceCounts));
+                    this.writeCSV(outputFolder + "/difference.csv",
```
difference.csv -> runSummaryDifference.csv

danielduhh commented on src/main/java/org/openstreetmap/atlas/checks/commands/FlagStatisticsSubCommand.java:
```diff
@@ -0,0 +1,619 @@
+            // Write the counts output if requested
+            if (outputTypes.contains("counts"))
+            {
+                this.writeCSV(outputFolder + "/counts.csv", countsOutput);
```
Can you think of something more informative that `counts.csv`?

danielduhh commented on src/main/java/org/openstreetmap/atlas/checks/commands/FlagStatisticsSubCommand.java:
```diff
@@ -0,0 +1,619 @@
+     *            {@link String}
+     * @return a 2D {@link List} of {@link String}s
+     */
+    private List<List<String>> generateTotalsOutputStage2(final List<List<String>> stage1Output,
```
This will require updating the name of the "stage1" function as well...

danielduhh commented on src/main/java/org/openstreetmap/atlas/checks/commands/FlagStatisticsSubCommand.java:
```diff
@@ -0,0 +1,619 @@
+     *            {@link String}
+     * @return a 2D {@link List} of {@link String}s
+     */
+    private List<List<String>> generateCountsOutputStage2(final List<List<String>> stage1Output,
```
Same comment about stage1/stage2

danielduhh commented on src/main/java/org/openstreetmap/atlas/checks/commands/FlagStatisticsSubCommand.java:
```diff
@@ -0,0 +1,619 @@
+        final List<List<String>> outputLines = new ArrayList<>(stage1Output);
+
+        // Add new headers
+        outputLines.get(0).add(1, "Reference");
```
How do you feel about indicating these fields are sum's in the headers. Maybe  `Sum(Reference)` and `Sum(Difference)` ?

danielduhh commented on src/main/java/org/openstreetmap/atlas/checks/commands/FlagStatisticsSubCommand.java:
```diff
@@ -111,32 +116,30 @@ public int execute()
             if (outputTypes.contains("totals"))
             {
-                this.writeCSV(outputFolder + "/totals.csv", totalsOutput);
+                this.writeCSV(outputFolder + "/checkSummary.csv", totalsOutput);
```
Should we change the output type name as well? Maybe we can put all of the supported `output-types` in a enum set

Bentleysb replied:
Done
